### PR TITLE
[qat] Quantization aware training in eager mode

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #22899 [qat] Conv module&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D16274792/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#22732 [qat] Quantization aware training in eager mode**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D16199356/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #22714 [qat] Add nn.qat.Linear&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D16193585/)

Add support for quantization aware training in eager mode

Modifications to Post training flow:
## Prepare
* Fusion: e.g. (Conv, Bn) → ConvBn (float)
* Swapping: To insert fake_quant to weight, we need to swap the float modules that has weight with different qat modules, e.g. Conv → torch.nn.qat.Conv , ConvBn → torch.nn._intrinsic.qat.ConvBn
```
    * previously we were thinking about modify the weight in forward_pre hook and change it back in forward_hook:
        * def forward_pre_hook(self, input):
                self.float_weight = self.weight
                self.weight = self.fake_quantize(self.float_weight)

            def forward_hook(self, input):
                self.weight = self.float_weight
```

* Assignments to self.weight are needed because we can’t change forward function and in forward function they are using self.weight.
* But we will need to keep two copies of weight in this case, so it’s probably better to just swap the module
* So we want to just swap Conv to torch.nn.qat.Conv and Linear to torch.nn.qat.Linear
* qat modules will have fake_quant for output and weights inserted in forward function

## Convert
* flow should be identical to ptq, but the swapping dictionary is slightly different since modules are changed in prepare step.

Differential Revision: [D16199356](https://our.internmc.facebook.com/intern/diff/D16199356/)

### CODE REVIEW

File: torch/nn/quantized/modules/linear.py [hunk 1]
```diff
@@ -150,13 +151,18 @@ def from_float(mod):
             Args: `mod` a float module, either produced by torch.quantization utilities
             or directly from user
         """
-        assert type(mod) == NNLinear, 'nnq.Linear.from_float only works for nn.Linear'
-        assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
-        assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
-        activation_observer = mod.observer
+        if mod.training:
```
Comment (dzhulgakov) [inline torch/nn/quantized/modules/linear.py]: it's a bit weird to rely here on training/eval, isn't it? maybe just check what it was instrumented with? (i.e. whether it's weight_fake_quant or not

File: torch/quantization/__init__.py [hunk 2]
```diff
@@ -12,6 +13,23 @@ def default_eval_fn(model, calib_data):
     for data in calib_data:
         model(data)
 
+default_loss_fn = torch.nn.MSELoss(reduction='sum')
+def default_train_fn(model, train_data, loss_fn=default_loss_fn, optimizer=None):
+    r"""
+    Default train function takes a torch.utils.data.Dataset and train the model
+    on the dataset
+    """
+    if not optimizer:
```
Comment (dzhulgakov) [inline torch/quantization/__init__.py]: it's pretty sketchy to do it

File: torch/quantization/__init__.py [hunk 3]
```diff
@@ -12,6 +13,23 @@ def default_eval_fn(model, calib_data):
...
+        optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
+    for i in range(10):
```
Comment (dzhulgakov) [inline torch/quantization/__init__.py]: this is absolutely random - why do we have it? and I don't think we should provide a training loop in this case at all

File: torch/quantization/quantize.py [hunk 4]
```diff
@@ -92,9 +93,10 @@ class QuantWrapper(nn.Module):
     def __init__(self, module):
         super(QuantWrapper, self).__init__()
         qconfig = module.qconfig if hasattr(module, 'qconfig') else None
-        self.quant = QuantStub(qconfig)
-        self.dequant = DeQuantStub()
-        self.module = module
+        self.add_module('quant', QuantStub(qconfig))
+        self.add_module('dequant', DeQuantStub())
+        self.add_module('module', module)
+        self.train(module.training)
```
Comment (dzhulgakov) [inline torch/quantization/quantize.py]: nit: self.training = module.training

File: torch/quantization/quantize.py [hunk 5]
```diff
@@ -123,25 +125,27 @@ def add_quant_dequant(module):
...
     """
-    propagate_qconfig(module, qconfig_dict)
+    propagate_qconfig(model, qconfig_dict)
     if qconfig_dict:
-        module = add_quant_dequant(module)
-    add_observer(module)
-    return module
+        model = add_quant_dequant(model)
+    add_observer(model)
+    if model.training:
```
Comment (dzhulgakov) [inline torch/quantization/quantize.py]: it feels to fragile to rely on it. Top-level different functions for qat and pqt feel better. In the end, you'd still need to pass very different "eval" function into them

File: torch/quantization/quantize.py [hunk 6]
```diff
@@ -247,7 +258,7 @@ def swap_module(mod, mapping):
         if type(mod) in mapping:
             new_mod = mapping[type(mod)].from_float(mod)
 
-    if type(mod) == DeQuantStub:
+    if swap_dequant and type(mod) == DeQuantStub:
```
Comment (dzhulgakov) [inline torch/quantization/quantize.py]: this is sketchy - maybe instead just say `if DequantStub in mapping`? (so the only difference is that we ignore the observer field)

File: torch/nn/quantized/modules/linear.py (hunk 1 above)
Comment (jerryzh168) [reply torch/nn/quantized/modules/linear.py]: why? we are already using this to control the overall flow

File: torch/quantization/__init__.py (hunk 3 above)
Comment (jerryzh168) [reply torch/quantization/__init__.py]: I can move this to the test file

File: torch/quantization/quantize.py (hunk 4 above)
Comment (jerryzh168) [reply torch/quantization/quantize.py]: using this because we want to change the submodules as well

Comment (jerryzh168) [reply torch/quantization/quantize.py]: specifically: QuantStub and DeQuantStub, although we didn't rely on this right now

File: torch/quantization/quantize.py (hunk 5 above)
Comment (jerryzh168) [reply torch/quantization/quantize.py]: I thought we want a one line API for both, no?
I guess the reason we have eval_fn in the argument of top level API is to allow unification and this is applicable for graph mode as well

File: torch/quantization/quantize.py (hunk 6 above)
Comment (jerryzh168) [reply torch/quantization/quantize.py]: I can add qconfig to DeQuantStub and remove this argument

File: torch/quantization/__init__.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/__init__.py]: removed, and the function is moved to common_quantization.py

File: torch/nn/quantized/modules/linear.py (hunk 1 above)
Comment (raghuramank100) [reply torch/nn/quantized/modules/linear.py]: @jerryzh168 : What if we just wanted to insert fake-quants for weights to simulate accuracy for eval? Will this be too restrictive in that case?

File: torch/quantization/__init__.py (hunk 3 above)
Comment (raghuramank100) [reply torch/quantization/__init__.py]: Fully agree, we do not need this.

File: torch/quantization/fake_quantize.py [hunk 7]
```diff
@@ -43,3 +44,16 @@ def forward(self, X):
                 X, self.scale.double(), self.zero_point.long(), self.quant_min,
                 self.quant_max)
         return X
+
```
Comment (raghuramank100) [inline torch/quantization/fake_quantize.py]: Comment on line 39: Can we remove if self.training: just self.enable_fq should be sufficient. This allows us to use fake quant in eval to model quantization errors.

File: torch/quantization/quantize.py (hunk 5 above)
Comment (raghuramank100) [reply torch/quantization/quantize.py]: For prepare, the changes required to support QAT and PTQ are small. Why not try to have the same API for QAT and PTQ?.

File: torch/quantization/quantize.py [hunk 8]
```diff
@@ -169,46 +173,53 @@ def __init__(self):
...
     Return:
-        A quantized module
+        A quantized model
     """
-    module = prepare(module, qconfig_dict)
-    eval_fn(module, eval_args)
-    convert(module)
-    return module
+    model = prepare(model, qconfig_dict)
+    eval_fn(model, eval_args)
```
Comment (raghuramank100) [inline torch/quantization/quantize.py]: Lets call this run_model instead of eval or training so that it is generic for both flows.

File: torch/quantization/quantize.py [hunk 9]
```diff
@@ -123,25 +125,27 @@ def add_quant_dequant(module):
...
         training
     """
-    propagate_qconfig(module, qconfig_dict)
+    propagate_qconfig(model, qconfig_dict)
     if qconfig_dict:
-        module = add_quant_dequant(module)
-    add_observer(module)
-    return module
+        model = add_quant_dequant(model)
+    add_observer(model)
```
Comment (raghuramank100) [inline torch/quantization/quantize.py]: Lets call it as add_quantization_module(model) instead. Calling it observer is confusing.

File: test/test_quantization.py [hunk 10]
```diff
@@ -259,9 +260,29 @@ def checkQuantized(model):
...
+        model = quantize(ManualQuantModel().eval(), default_eval_fn, calib_data)
         checkQuantized(model)
 
+class QuantizationAwareTrainingTest(QuantizationTestCase):
+    def test_manual(self):
+        model = ManualQATModel()
+        model.qconfig = default_qat_qconfig
+
+        prepare(model)
+        self.checkObservers(model)
```
Comment (raghuramank100) [inline test/test_quantization.py]: Lets call it self.checkFakeQuant here so that it is clear.

File: test/test_quantization.py [hunk 11]
```diff
@@ -259,9 +260,29 @@ def checkQuantized(model):
...
+
+        default_eval_fn(model, calib_data)
```
Comment (raghuramank100) [inline test/test_quantization.py]: Should this be default_train_fn?

File: test/test_quantization.py [hunk 12]
```diff
@@ -259,9 +260,29 @@ def checkQuantized(model):
...
+
+        def checkQuantized(model):
+            self.assertEqual(type(model.fc1), nnq.Linear)
+            self.assertEqual(type(model.fc2), nnq.Linear)
+            default_eval_fn(model, calib_data)
+
+        model = ManualQATModel()
+        model.qconfig = default_qat_qconfig
+        model = quantize(model, default_train_fn, train_data)
+        checkQuantized(model)
```
Comment (raghuramank100) [inline test/test_quantization.py]: Can we also have a test that allows for insertion of fake-quant on a model setup to be in evaluation mode?.

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please address comments, thanks

File: test/test_quantization.py (hunk 10 above)
Comment (jerryzh168) [reply test/test_quantization.py]: we need a unified name, this is for both ptq and qat.

File: test/test_quantization.py (hunk 11 above)
Comment (jerryzh168) [reply test/test_quantization.py]: oh right

File: torch/nn/quantized/modules/linear.py (hunk 1 above)
Comment (jerryzh168) [reply torch/nn/quantized/modules/linear.py]: not for this place in particular, this is called in last convert step, if we just insert fake-quant we won't hit this code.

File: torch/quantization/fake_quantize.py (hunk 7 above)
Comment (jerryzh168) [reply torch/quantization/fake_quantize.py]: Sure

File: torch/quantization/quantize.py (hunk 9 above)
Comment (jerryzh168) [reply torch/quantization/quantize.py]: we need a name to cover both observer and fake_quant

File: torch/quantization/quantize.py (hunk 8 above)
Comment (jerryzh168) [reply torch/quantization/quantize.py]: changing to `run_fn` and `run_args`

File: test/test_quantization.py (hunk 12 above)
Comment (jerryzh168) [reply test/test_quantization.py]: added a test, we just need to call prepare

File: torch/nn/quantized/modules/linear.py [hunk 13]
```diff
@@ -154,13 +155,18 @@ def from_float(mod):
             Args: `mod` a float module, either produced by torch.quantization utilities
             or directly from user
         """
-        assert type(mod) == NNLinear, 'nnq.Linear.from_float only works for nn.Linear'
-        assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
-        assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
-        activation_observer = mod.observer
+        if hasattr(mod, 'weight_fake_quant'):
+            # assert type(mod) == QATLinear, 'training mode nnq.Linear.from_float only works for nn.qat.Linear'
+            activation_observer = mod.observer
```
Comment (dzhulgakov) [inline torch/nn/quantized/modules/linear.py]: this can be pulled out outside of the "if" (happens in both branches)

File: torch/quantization/__init__.py [hunk 14]
```diff
@@ -3,14 +3,20 @@
...
     """
-    for data in calib_data:
-        model(data)
+    total, correct = 0, 0
+    for data, target in calib_data:
+        output = model(data)
+        _, predicted = torch.max(output, 1)
+        total += target.size(0)
+        correct += (predicted == target).sum().item()
+    return correct / total
```
Comment (dzhulgakov) [inline torch/quantization/__init__.py]: why do we have this as a default function? it feels very specialized to classification problem

File: torch/quantization/quantize.py (hunk 5 above)
Comment (dzhulgakov) [reply torch/quantization/quantize.py]: I don't feel strongly about it. But that's why it feels weird - the qconfig one passes is going to be different for QAT and PTQ. So it's more logical for prepare() to just enforce (or switch) model to be in the right eval/train state. In this case we have two switches which need to work together.

File: torch/quantization/__init__.py (hunk 14 above)
Comment (jerryzh168) [reply torch/quantization/__init__.py]: I can rename that to test_only.. we just need some function to run the tests.

File: torch/quantization/quantize.py (hunk 5 above)
Comment (jerryzh168) [reply torch/quantization/quantize.py]: qconfig pass is the same, we just pass in different qconfig_dict.

Comment (jerryzh168) [reply torch/quantization/quantize.py]: do you mean change API to `prepare(model, training=False, qconfig_dict=None)`?

File: torch/quantization/quantize.py (hunk 6 above)
Comment (jerryzh168) [reply torch/quantization/quantize.py]: why this is still here? I think we don't have the most recent code..

File: torch/quantization/quantize.py (hunk 5 above)
Comment (jerryzh168) [reply torch/quantization/quantize.py]: I don't know how this extra boolean is different than just asking user to always specify whether the model is in train mode or inference mode before calling quantize

Comment (jerryzh168) [reply torch/quantization/quantize.py]: will split the APIs.

File: test/test_quantization.py (hunk 10 above)
Comment (jerryzh168) [reply test/test_quantization.py]: plan to just call it observer for now

File: torch/quantization/quantize.py [hunk 15]
```diff
@@ -76,7 +81,7 @@ def add_observer(module):
     if hasattr(module, 'qconfig') and module.qconfig is not None and len(module._modules) == 0:
         # observer and hook will be gone after we swap the module
         module.add_module('observer', module.qconfig.activation())
-        module.register_forward_hook(_observer_forward_hook)
+        module.register_forward_hook(_fake_quant_forward_hook if module.training else _observer_forward_hook)
```
Comment (jerryzh168) [inline torch/quantization/quantize.py]: @dzhulgakov btw we still rely on module.training here, let me know if you want me to change this

Comment (dzhulgakov) [reply torch/quantization/quantize.py]: @jerryzh168 - I don't think you need it. You can just have a single hook that does `return observer(output)`. In case of actual observer, it returns None so the output won't be overwritten. Effectively we pass down the contract of forward hooks down to observer (we should document it). Want to fix it?

Comment (jerryzh168) [reply torch/quantization/quantize.py]: sounds hacky, should we do that?

Comment (dzhulgakov) [reply torch/quantization/quantize.py]: I'd say it's less magical than checking .training here. Btw, if .training is wrong, e.g. it gets to be 'true' for PQT, you'd do exactly what I've suggested :)

Comment (jerryzh168) [reply torch/quantization/quantize.py]: we are putting a constraint on observer that it will always return None in forward and on FakeQuant module that it will return a valid output in forward though. 

The constraint for observer will be relaxed in current implementation, but we require people to get the training flag right, which is another constraint. when .training is wrong the FakeQuant module will just be an observer. which still make sense actually.

I don't have much preference in this case. we can use the original one.
